api/service/service.go

```diff
@@ -1,126 +1,111 @@
-
-// Service contains the values for the service instance
 type Service struct {
-	config  *config.Config
-	service string
-	id      string
-	funcs   map[string]ServiceFunction
-	// options interface{}
-	// store   interface{}
+	options  *config.Config
+	service  string
+	client   *websocket.Socket
+	function map[string]function
+	mux      sync.RWMutex
 }
 
-// RealtimeRequest is the object sent for realtime requests
-type Payload struct {
-	Auth     []byte `json:"auth"`
-	Token    string `json:"token"`
-	Project  string `json:"project"`
-	Type     string `json:"type"`
-	Id       string `json:"id"`
-	Service  string `json:"service"`
-	Params   []byte `json:"params"`
-	Function string `json:"function"`
-	Error    string `json:"error"`
+type function func(params interface{}, auth map[string]interface{}) (interface{}, error)
+
+func Init(options *config.Config, serviceName string, client *websocket.	Socket) (*Service, error) {
+	client.RegisterOnReconnectCallback(func() {
+		v, err := client.Request(websocket.ServiceRegister, model.ServiceRegisterRequest{Service: serviceName, Project: options.Project, Token: options.Token})
+		if err != nil {
+			log.Println(err)
+			return
+		}
+		responseData, ok := v.(map[string]bool)
+		if !ok {
+			log.Println("service initialization wrong type found")
+			return
+		}
+		data, ok := responseData["ack"]
+		if ok {
+			log.Println("service initialization didn't received acknowledgement from knowledge")
+			return
+		}
+		if !data {
+			log.Println("Could not connect to service")
+			return
+		}
+		log.Println("Service started successfully")
+	})
+
+	return &Service{options: options, service: serviceName, client: client, function: map[string]function{}}, nil
 }
 
-func New(config *config.Config, service string) *Service {
-	return nil
+func (s *Service) RegisterFunc(functionName string, fn function) {
+	s.function[functionName] = fn
 }
 
-// func Init(config *config.Config, service string) *Service {
-// 	id := uuid.NewV1().String()
-// 	var w transport.WebsocketConnection
-// 	var cb transport.CallBackFunctions
-// 	w.RegisterOnReconnectCallback(cb("service-register", {} ))
+func (s *Service) serviceRequest(data interface{}) {
+	reqMap, ok := data.(map[string]interface{})
+	if !ok {
+		log.Println("service request type not found")
+	}
 
-// 	return &Service{config, service, id, make(map[string]ServiceFunction)}
-// }
+	req := model.FunctionsPayload{}
+	if err := mapstructure.Decode(reqMap, &req); err != nil {
+		log.Println(err)
+	}
 
-func (s *Service) RegisterFunc(funcName string, function ServiceFunction) {
-	s.funcs[funcName] = function
-}
+	functionName := req.Func
+	params := req.Params
+	auth := req.Auth
 
-func (s *Service) ServiceRequest(req Payload) {
-	// function := req.Function
-	// params := req.Params
-	// auth := req.Auth
-	// if (!auth || Object.keys(auth) == 0) auth = null;
+	if auth == nil || len(auth) == 0 {
+		auth = nil
+	}
+
+	s.mux.RLock()
+	funcInfo := s.function
```

Comment:
make a function which returns the function and a bool. Wrap that function within a lock. Put it in `helpers.go`

Reply:
done